src/cpu/o3/fetch.cc

```diff
@@ -1245,6 +1245,14 @@ Fetch::fetch(bool &status_change)
                     staticInst = curMacroop->fetchMicroop(this_pc.microPC());
                 }
                 newMacro |= staticInst->isLastMicroop();
+
+                /// TODO: Make it proper
+                // This is a hack to make sure that the size of the macroop
+                // is propergated to the microops. This is needed for
+                // for the branch predictor to compute the return address
+                // in the decoupled frontend.
+                if (curMacroop)
```

Comment:
This seems slightly ugly since it would mean that adding up the size of the micro ops within a macro op would result in a size that is larger than the macro op. I guess we could use a convention where the macro op has a size and the micro ops reflect that size (which is what you are implementing), but that seems like it might cause issues down the line.

Is there any chance we could track this in the branch predictor instead? And should we even need perform predictions on micro ops?

Reply:
I am open to any other suggestions on how to handle that.
The problem is the following: You have a branch instruction that gets broken into let's say the address calc and the actual jump microop. (All have the same instruction address but different microPCs) The BPU will be queried only for the jump micro-op.
When the prediction is fall-through the BPU computes the next InstAddr. For this, it needs the size of the overall macroop.
At the moment this is done by advancing the PCState. This state knows when and where the instruction ends.
For a decoupled fetch PC this is not available as you search instruction addresses until you hit in the BTB and you don't have a valid PC state.

This is easy for RISC-V or ARM but we also want to support x86

Reply:
IMHO, reflecting the instruction size in each micro-op is not wrong since they are just conceptual and have no size. They are microarchitecture-dependent

Reply:
I like the solution @dhschall proposes here. It's not ideal, but I don't think there are any other solutions that I can see.

Reply:
Oops, spoke too soon. It would still be nice if this was in the constructor of the micro-op instead of having to be replicated in all CPU models.

Reply:
That sounds like a good reason to keep the size in the micro-op. It's still slightly weird for micro-code sequences where the with a branch that isn't the last micro-op. I don't think that's very common though, so we probably don't need to worry about that at the moment.

I agree with Jason that this should be handled in the constructor rather than the CPU model unless there is a really good reason.

Reply:
Thanks for the suggestions. I think I found a way to easily pass the instruction size of all micro-ops.
Let me know if this is ok.

Reply:
Thanks, this looks like a better solution to me.